### DISCUSSION TITLE
Better way to iterate over messages / collections

### DISCUSSION THREAD

**Discussion by muhcdan** [nextgenhealthcare/connect#5861] [General]
I realized something recently, we are generally using for each statements rather that for statements for iterating hl7 messages.  Why is that? for each is long deprecated from ecma 4.  Isn't there a better way to iterate?

```
// e.g current way:
for each (mrn in hl7AsXML['PID']['PID.3']) {
}

// doesn't work?
for  (mrn in hl7AsXML['PID']['PID.3']) {
}

```

**Comment by pacmano1** [comment]
I don't understand your question.  ECMA-4 was never adopted.

And Mirth uses e4x, an extension to ECMA to support XML.

And the syntax is typically:

```JS
for each (mrn in msg['PID']['PID.3']) {
	logger.info(mrn['PID.3.1'].toString())
}
```

If you want to use a for, then you need to do that with a length.
```JS
for (var i = 0; i < msg['PID']['PID.3'].length; i++) {
	logger.info(msg['PID']['PID.3'][i]['PID.3.1'].toString())	
}
```

There are advantages to using for each, including when you delete elements.

**Reply by muhcdan** [reply]
Right ecma4 was never adopted and "for each" is ecma4 and is deprecated.  It's considered a synatax error with ecma5 and 6. 

When I use mirthsync the IDE shows me syntax errors when it sees "for each".   The only work-around I found was to set the .js to be "ActionScript". 

what does 

for (mrn in msg['PID']['PID.3']) {
	logger.info(mrn['PID.3.1'].toString())
}

do differently? I only see nulls.

**Reply by pacmano1** [reply]
Post a sample message where the code I posted returns nulls.
